client/src/main/java/io/pravega/client/connection/impl/IoBuffer.java

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.connection.impl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.pravega.common.MathHelpers;
+import io.pravega.shared.protocol.netty.WireCommands;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+class IoBuffer {
+    private int maxBufferSize = WireCommands.MAX_WIRECOMMAND_SIZE;
+    private ByteBuffer buffer = null;
+    
+    private ByteBuf sliceOut(int size) {
+        ByteBuf result = Unpooled.wrappedBuffer(buffer.array(), buffer.arrayOffset() + buffer.position(), size);
+        buffer.position(buffer.position() + size);
+        if (!buffer.hasRemaining()) {
+            buffer = null;
+        }
+        return result;
+    }
+    
+    public ByteBuf getBuffOfSize(InputStream in, int size) throws IOException {
```

Comment:
Added a comment.

> ByteBuf returned can have a reference to two byte-arrays (at max) where the size of each byte-array can be at max 16MB

Each can be at most 8MB. In practice however one is going to be small and the other big. Because they are allocated according to what is available (which is bounded by `TcpClientConnection.TCP_BUFFER_SIZE = 256kb` ) and then what is requested. Because the request is bounded by 8mb this means that the 'tail' of the first buffer plus the whole of the second buffer is <= 8mb. 

> Both the byte arrays will be freed only when the ByteBuf.release is invoked. And in case of small "size" params all the byteBufs should invoke release before the byte arrays are freed.

It's actually different depending on how much data is read. If the amount read out is smaller than the size of the buffer a view of it is returned and the array is still held by this class. This means that even once release is called the buffer will not be freed. (Which is good because it still has unread data)

In the event that the read is larger than the buffer the whole of the buffer is passed into the returned buffer and the IoBuffer class does not hold onto it. Similarly the remaining bytes are passed in the returned buffer and not held elsewhere. In this case because the only reference to the data is inside the ByteBuf, it actually does not matter if `free` is called or not, because once the ByteBuf is dropped there will be no references to the data anywhere and it can be garbage collected normally.

So while ByteBuf is designed to let you pool and re-use memory by using reference counting, this code is not actually taking advantage of that feature and is just relying on normal garbage collection. If that were to change in the future, the code would still be correct because the code that receives the ByteBufs still calls release (according to our unit tests) so it would be OK to add such functionality.